Microsoft.AVS.VMFS/Microsoft.AVS.VMFS.psm1

```diff
@@ -568,3 +568,286 @@ function Remove-VMHostStaticIScsiTargets {
+                            $Name, $AdminQueueSize, $ControllerId, 
+                            $null, $IoQueueNumber, $IoQueueSize, $NodeAddress,
+                            $KeepAliveTimeout, $PortNumber, $StorageSystemNQN, $null, $null 
+                        );
+       
+                        if ($EsxCliResult) {
+                            Write-Host "ESXi host $($VmHost.Name) is connected to storage controller via " $Name
+                        }
+                        else {
+                            Write-Host "Failed to connect ESXi host $($VmHost.Name) to storage controller "
+                        }
+            
+                        Write-Host "Connecting Controller status: "$EsxCliResult;
+                    }
+                    catch {
+                        Write-Error "Failed to connect ESXi NVMe/TCP storage adapter to storage controller. $($_.Exception) " 
+                    }
+                }
+            }
+        } 
+        Write-Host "Rescanning NVMe/TCP storage adapter.."
+        $RescanResult = Get-VMHostStorage -VMHost $VmHost.Name -RescanAllHba 
+        Write-Host "Rescanning Completed."
+        Write-Host ""
+    }
+  
+  
+} 
+ 
+<#
+    .SYNOPSIS
+     This function disconnects all ESXi host(s) from the specified storage cluster node/target.
+
+     1. vSphere Cluster Name
+     2. Storage SystemNQN
+
+    .PARAMETER ClusterName
+     vSphere Cluster Name
+    
+    .PARAMETER StorageSystemNQN
+     Storage system NQN
+
+    .EXAMPLE
+     Disconnect-NVMeTCPTarget -ClusterName "Cluster-001"  -StorageSystemNQN "nqn.2016-01.com.lightbitslabs:uuid:46edb489-ba18-4dd4-a157-1d8eb8c32e21"
+
+    .INPUTS
+     vSphere Cluster Name, Storage SystemNQN
+
+    .OUTPUTS
+     None.
+#>
+function Disconnect-NVMeTCPTarget {
+    [CmdletBinding()]
+    [AVSAttribute(10, UpdatesSDDC = $false)]
+   
+    Param
+    (
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'vSphere Cluster Name')]
+        [String] $ClusterName,
+                
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Target storage SystemNQN')]
+        [string] $StorageSystemNQN
+    )
+
+    Write-Host "Disconnecting ESXi host(s) from storage target under Cluster " $ClusterName
+    Write-Host " " ;
+
+    $Cluster = Get-Cluster -Name $ClusterName -ErrorAction Ignore
+    if (-not $Cluster) {
+        throw "Cluster $ClusterName does not exist."
+    }
+
+    $VmHosts = $Cluster | Get-VMHost
+    foreach ($VmHost in $VMHosts) {
+
+        if ($VmHost.ConnectionState -ne "Connected") {
+            Write-Host "ESXi host $($VmHost.Name)  must be in connected state, ignoring operation because of current state "$VmHost.ConnectionState
+            Write-Host ""
+            continue
+        }
+
+        $ProvisionedDevices = Get-Datastore -VMHost $VmHost.Name | where-object{$_.ExtensionData.Info.Vmfs.Extent.DiskName -like  'eui.*'}
```

Comment:
@khensler 
Added code to skip disconnecting ESXi host(s) that has at least one  provisioned device from target.